src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,148 @@
+  auto& probe_chunk = input_table->get_chunk(ChunkID{0});
+  const auto& probe_segment = probe_chunk.get_segment(ColumnID{0});
+  auto probe_ref_segment = std::dynamic_pointer_cast<ReferenceSegment>(probe_segment);
+  if (probe_ref_segment) {
+    input_table = probe_ref_segment->referenced_table();
+    for (auto position : *(probe_ref_segment->pos_list())) {
+      auto chunk_id = position.chunk_id;
+      auto chunk_offset = position.chunk_offset;
+
+      auto segment = input_table->get_chunk(chunk_id).get_segment(_column_id);
+      resolve_data_type(col_type, [&](const auto data_type_t) {
+        using Type = typename decltype(data_type_t)::type;
+        auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+        auto _comparator = _get_comparator_method<Type>();
+        if (value_segment) {
+          // Value Segment
+          auto comparison_value = type_cast<Type>(value_segment->values()[chunk_offset]);
+          if (_comparator(type_cast<Type>(_search_value), comparison_value)) {
+            pos_list->push_back(RowID{chunk_id, chunk_offset});
+          }
+        } else {
+          // Dictionary Segment
+          auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+          auto comparison_value = type_cast<Type>(dictionary_segment->get(chunk_offset));
+          if (_comparator(type_cast<Type>(_search_value), comparison_value)) {
+            pos_list->push_back(RowID{chunk_id, chunk_offset});
+          }
+        }
+      });
+    }
+  } else {
+    auto chunk_count = input_table->chunk_count();
+    for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+      auto& chunk = input_table->get_chunk(chunk_id);
+      std::shared_ptr<BaseSegment> segment = chunk.get_segment(_column_id);
+
+      // Fill position list for different segment types
+      resolve_data_type(col_type, [&](const auto data_type_t) {
+        using Type = typename decltype(data_type_t)::type;
+        auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+        auto _comparator = _get_comparator_method<Type>();
+        if (value_segment) {
+          // Value Segment
+          auto chunk_size = chunk.size();
+          for (auto chunk_offset = ChunkOffset{0}; chunk_offset < chunk_size; ++chunk_offset) {
+            // TODO(anyone): extract method
+            auto comparison_value = type_cast<Type>(value_segment->values()[chunk_offset]);
+            if (_comparator(type_cast<Type>(_search_value), comparison_value)) {
+              pos_list->push_back(RowID{chunk_id, chunk_offset});
+            }
+          }
+        } else {
+          // Dictionary Segment
+          auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+          auto chunk_size = chunk.size();
+          for (auto chunk_offset = ChunkOffset{0}; chunk_offset < chunk_size; ++chunk_offset) {
+            // TODO(anyone): extract method
```

Comment:
Is this still wanted? made an issue for that in case.
```suggestion
            // TODO(anyone): extract method (#48)
```